src/Avalonia.Visuals/Visual.cs

```diff
@@ -328,6 +328,44 @@ protected static void AffectsRender(params AvaloniaProperty[] properties)
             }
         }
 
+        /// <summary>
+        /// Indicates that a brush property change should cause <see cref="InvalidateVisual"/> to be
+        /// called.
+        /// </summary>
+        /// <param name="properties">The properties.</param>
+        /// <remarks>
+        /// This method should be called in a control's static constructor with each property
+        /// on the control which when changed should cause a redraw. It not only triggers an
+        /// invalidation when the property itself changes, but also when the brush raises
+        /// the <see cref="IMutableBrush.Changed"/> event.
+        /// </remarks>
+        protected static void BrushAffectsRender<T>(params AvaloniaProperty<IBrush>[] properties)
```

Comment:
I think it might be better to introduce an interface like `IAffectsRender` (with a Changed event) that when implemented by a DP, the `AffectsRender` method automatically hooks into that event. That way the API remains agnostic to the fact that it's a brush. `Geometry` could also implement it, for example.

Reply:
The `property.Changed` event is untyped anyhow. So why not just do the same thing you did here in the `AffectsRender` method? Only use a generic interface instead of `IMutableBrush`. For example:

```csharp
if (e.OldValue is IAffectsRender oldValue)
{
    oldValue.Changed -= ...
}
```

The only downside is that each value change for a property marked as AffectsRender would have two additional casts during assignment.

But the upside is that the API surface is smaller and less error-prone.

Reply:
Ah, yes I missed the fact that you wanted to do it automatically in `AffectsRender`. We don't want to do that. For example, `Button` has a `Background` property but when that changes we don't want to call `InvalidateVisual` because `Button` doesn't actually render anything: the actual rendering is done by the `Border` or whatever in the `Button`'s template. I mentioned this in the PR body:

> Note that only controls that do their own rendering need to call BrushAffectsRender - templated controls have no visual representation of their own so they don't need to call it.

So I think it this behaviour needs to be opted-in.

This will become even more vital once we extend this pattern to `Geometry` as changing that can cause a remeasure and we don't want to queue more remeasures than necessary.

Reply:
Yeah, but `Button` wouldn't have `AffectsRender` for background, right? Haven't you changed every `AffectsRender` that had a brush with it to `BrushAffectsRender`? If a property does affect the render, then also changes to its internal structure should. Same goes for geometries.

Reply:
Ah yes, you're right of course. I seem to have had a brainfart ;) Sorry, that took me way to long to understand what you were saying.

Reply:
Ok, make that change!